### PR TITLE
Fix Measuring Child AMP Elements

### LINKED ISSUE #9267: <amp-img> children of <amp-selector> don't load without user intervention
## What's the issue?

When an `amp-selector` container contains `amp-img` children, the images do not load until the user interacts with the page, for example with a tap, or by resizing the browser on desktop.

Furthermore, if the `amp-accordion` script is included in document head, even though it is not being used, then the problem does not occur, i.e. the images are loaded and rendered on page load.

## How to reproduce the issue?

See jsbin here: [http://jsbin.com/lumuvonanu/edit?html,output](http://jsbin.com/lumuvonanu/edit?html,output)

1. Load jsbin above
2. Observe images not loading
3. Resize browser
4. Observe images load and render
<br><br>
5. Reload jsbin above
6. Uncomment amp-accordion script so that it's included in page
7. Rerun jsbin 
8. Observe that image load and render

<br><br>
Note that the code is copied from this ampbyexample page, which contains the `amp-accordion` script, and so the issue does not occur on this page
https://ampbyexample.com/components/amp-selector/

## What browsers are affected?
Desktop: Chrome, Safari, Firefox, Opera on OSX
Mobile: Chrome, Opera, Firefox on Android

## Expected
1. `amp-img` inside `amp-selector `should load on page load
2. Inclusion of `amp-accordion` script should not affect the loading of any `amp-img` when no `amp-accordion` is used on page.

## Full code to reproduce
```
<!doctype html>
<html ⚡>
<head>
  <meta charset="utf-8">
  <link rel="canonical" href="https://ampbyexample.com/components/amp-selector/">
  <meta name="viewport" content="width=device-width,minimum-scale=1,initial-scale=1">
  <script async src="https://cdn.ampproject.org/v0.js"></script>
  <!-- ## Setup -->
  <!--
    Import the amp-selector component.
  -->
  <script async custom-element="amp-selector" src="https://cdn.ampproject.org/v0/amp-selector-0.1.js"></script>
  <script async custom-element="amp-form" src="https://cdn.ampproject.org/v0/amp-form-0.1.js"></script>
<!-- <script async custom-element="amp-accordion" src="https://cdn.ampproject.org/v0/amp-accordion-0.1.js"></script> -->
  <style amp-boilerplate>body{-webkit-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-moz-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-ms-animation:-amp-start 8s steps(1,end) 0s 1 normal both;animation:-amp-start 8s steps(1,end) 0s 1 normal both}@-webkit-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-moz-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-ms-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-o-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}</style><noscript><style amp-boilerplate>body{-webkit-animation:none;-moz-animation:none;-ms-animation:none;animation:none}</style></noscript>
<style amp-custom>
amp-selector:not([disabled]) [option][selected]:not([disabled]) {
  outline-color: #b60845;
}
/* disables build-in browser highlights */
amp-selector *:focus {
  outline: none;
}
amp-selector[disabled],
amp-selector [disabled] {
  opacity: 0.5;
}
.divider {
  border-right: 2px solid #333;
  height: 40px;
}
</style>
</head>
<body>
  The <code>amp-img</code> images below should load when the page first page loads, but they don't, unless you interact with the page, for example if you
  <ol>
    <li>You resize browser/rotate device, OR</li>
    <li>Uncomment the amp-accordion script (which is otherwise unused) in the document head</li>
  </ol>
  
  <amp-selector layout="container">
    <amp-img src="https://ampbyexample.com/img/landscape_sea_300x199.jpg"  width="60" height="40" option="1"></amp-img>
    <amp-img src="https://ampbyexample.com/img/landscape_desert_300x200.jpg" width="60" height="40" option="2"></amp-img>
    <div class="divider inline-block mx1"></div>
    <amp-img src="https://ampbyexample.com/img/landscape_ship_300x200.jpg" width="60" height="40" option="3"></amp-img>
    <amp-img src="https://ampbyexample.com/img/landscape_village_300x200.jpg" width="60" height="40" option="4"></amp-img>
  </amp-selector>

</body>
</html>
```

### PR DESCRIPTION
**Pull request by jridgewell**
AMP elements will not be measured until their ancestor AMP elements have been built. This excludes placeholders, which are laid out independent of their ancestor.

Fixes #9267, likely others.

### CODE REVIEW

**Comment by jridgewell** [comment]
Preview at http://output.jsbin.com/soniwap/quiet

File: css/amp.css [hunk 1]
```diff
@@ -271,7 +271,7 @@ i-amphtml-sizer {
   color: transparent !important;
 }
 
-.i-amphtml-notbuilt > * {
+.i-amphtml-notbuilt:not(.i-amphtml-layout-container ) > * {
```
Comment (dvoytenko) [inline css/amp.css]: nit: remove trailing space before `)`

File: src/service/resource.js [hunk 2]
```diff
@@ -365,19 +390,29 @@ export class Resource {
...
+      }
+      // If the current resource (not the ancestor) is a placeholder, allow it to
+      // measure. This is because placeholders are laid out independently of
+      // the ancestor element (even if they are sized and "owned" by it).
+      if (current.isPlaceholder_) {
+        break;
+      }
+      // If this ancestor isn't built yet, that means we don't know what kind
+      // of styles will be applied to this element. We must wait.
+      if (!ancestor.isBuilt()) {
```
Comment (dvoytenko) [inline src/service/resource.js]: This is only true for non-container ancestors. With containers, we want to measure as soon as we can. We non-containers, we already all good, right? It will measure to 0 and then re-measure to the right value.

File: src/service/resource.js [hunk 3]
```diff
@@ -147,6 +147,9 @@ export class Resource {
     /** @private {!AmpElement|undefined|null} */
     this.owner_ = undefined;
 
+    /** @private {!AmpElement|undefined|null} */
+    this.ampAncestor_ = undefined;
```
Comment (dvoytenko) [inline src/service/resource.js]: This needs to be discussed a bit more. But CSS changes seem rather clear to me. Can we merge them first and split this part into another PR? I'd like to get a bit deeper on this.

File: src/service/resource.js [hunk 4]
```diff
@@ -365,19 +390,29 @@ export class Resource {
...
-        // Use prefix to recognize AMP element. This is necessary because stub
-        // may not be attached yet.
-        startsWith(this.element.parentElement.tagName, 'AMP-') &&
-        !(RESOURCE_PROP_ in this.element.parentElement)) {
-      return;
+    while (ancestor) {
+      const resource = Resource.forElementOptional(ancestor);
+      // If there's an AMP ancestor that's still not stubbed (has a Resource),
+      // we don't know what to do yet.
+      if (!resource) {
```
Comment (dvoytenko) [inline src/service/resource.js]: I believe, unfortunately, we do have a very few `amp-` tags that are not custom elements. Could you pls double-check via validator? Though validator wouldn't give the full picture since some extensions might create them via `createElement`

File: src/service/resource.js (hunk 3 above)
Comment (jridgewell) [inline src/service/resource.js]: [Done](https://github.com/ampproject/amphtml/pull/9282).

File: src/service/resource.js (hunk 4 above)
Comment (jridgewell) [inline src/service/resource.js]: I don't understand.

Comment (dvoytenko) [inline src/service/resource.js]: I.e. https://github.com/ampproject/amphtml/blob/4435faeedc6585acd3bf6758e746b0fa388eb257/extensions/amp-sticky-ad/1.0/amp-sticky-ad.js#L96

This code is probably benign, since no one ads stuff into this element from what I can tell. But others might not be so lucky.

File: src/service/resource.js (hunk 2 above)
Comment (jridgewell) [inline src/service/resource.js]: > With containers, we want to measure as soon as we can

Fixed.

>  We non-containers, we already all good, right? It will measure to 0 and then re-measure to the right value.

Unfortunately not. If the non-container isn't built yet, it will measure to 0 and never re-remeasure (because it thinks it has a proper measure). That's the [bug](https://github.com/ampproject/amphtml/issues/9267).

File: src/service/resource.js (hunk 4 above)
Comment (jridgewell) [inline src/service/resource.js]: @zhouyx: Is that supposed to be publicly stylable?

Comment (zhouyx) [inline src/service/resource.js]: yes

Comment (jridgewell) [inline src/service/resource.js]: Reviewed, https://github.com/ampproject/amphtml/issues/9283. Will update so those aren't considered.

File: src/dom.js [hunk 5]
```diff
@@ -29,6 +30,20 @@ const HTML_ESCAPE_REGEX = /(&|<|>|"|'|`)/g;
...
 /**
+ * Determines if this element is an AMP element
+ * @param {!Element}
+ * @return {boolean}
+ */
+export function isAmpElement(element) {
+  const tag = element.tagName;
+  // Use prefix to recognize AMP element. This is necessary because stub
+  // may not be attached yet.
+  return startsWith(tag, 'AMP-') &&
```
Comment (zhouyx) [inline src/dom.js]: why is checking element.classList.contain('i-amphtml-element') not sufficient?

Comment (jridgewell) [inline src/dom.js]: From the [original comment](https://github.com/ampproject/amphtml/pull/9279/files#diff-9379e12fba31cf69d032da3c00ba0c8eL372)

> We can tell whether the parent has been stubbed by whether a resource has been attached to it.

That means we have to be able to tell regardless of where the element is in the CustomElement lifecycle. And an element has to have been stubbed (`#connectedCallback`, the first step) to have the `i-amphtml-element` class.

**Comment by aghassemi** [comment]
Thanks for the pull request. Since this PR has been stale for a while, I am closing it. Please feel free to reopen as needed.

**Comment by miladmhp** [comment]
@aghassemi I Have the same issue with AMP image loading, sometimes images loaded for a long time, or even I just have seen a white box instead of the image. And with just resize the page even for 1 px, all images have been load immediately, It seems to need a shock to load images:).
Please help me in solving this, I'm Confused.

**Comment by aghassemi** [comment]
@miladmhp Could you please file an issue with a repro Url?
